### PR TITLE
Fix layout bug and add tests

### PR DESCRIPTION
**Pull request by phschaad**
- [x] Adds tests using the testing framework Jest
- [x] Tests the integrated graph library
- [x] Adds some tests for the vertical layout (specifically, the ranking phase)
- [x] Fixes a few bugs detected through testing, including mishandling of self-loops
- [x] Ensure the vertical layout doesn't get stuck if it fails due to some reason

### CODE REVIEW

File: src/layouter/graphlib/algorithms/cycles.ts [hunk 1]
```diff
@@ -4,24 +4,32 @@ import { stronglyConnectedComponents } from './components';
...
-    constructor(nodes: NodeT[]) {
+    constructor(nodes: NodeT[], edges: [string, string][]) {
         this._nodes = new Set(nodes);
+        this._edges = new Set();
+        for (const e of edges)
+            this._edges.add(e.toString());
     }
 
     public get length(): number {
-        return this._nodes.size;
```
Comment (tbennun) [inline src/layouter/graphlib/algorithms/cycles.ts:13]: Please explain

File: tests/unit/layouter/graphlib/algorithms/cycles.test.ts [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+} from '../../../../../src/layouter/graphlib/algorithms/cycles';
+import { DiGraph } from '../../../../../src/layouter/graphlib/di_graph';
+
+function testFindSimpleCycle(): void {
+    const g = new DiGraph();
+
+    // Construct graph.
+    // 0    1
+    //      ↓
+    //    ┌→2
```
Comment (tbennun) [inline tests/unit/layouter/graphlib/algorithms/cycles.test.ts:13]: Nice ascii/Unicode

**Review by tbennun — COMMENTED** [review]
Looks great, I only have some questions and a test request. Great work with the testing infrastructure!

File: src/layouter/graphlib/algorithms/cycles.ts (hunk 1 above)
Comment (phschaad) [reply src/layouter/graphlib/algorithms/cycles.ts:13]: Added comment

File: src/layouter/graphlib/graph.ts [hunk 3]
```diff
@@ -121,7 +125,7 @@ export class Graph<NodeT, EdgeT> implements GraphI<NodeT, EdgeT> {
     }
 
     public numberOfEdges(): number {
-        return this.edges().length;
+        return this.edges().length / 2;
```
[the remark this answers is not in the document]
Comment (phschaad) [reply src/layouter/graphlib/graph.ts:130]: Added comment

File: tests/unit/layouter/state_machine/sm_layouter.test.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    expect(n5.rank).toBe(5);
+}
+
+describe('Test vertical state machine layout ranking', () => {
+    test('Basic branching', testBasicBranching);
+    test('Nested branching', testNestedBranching);
+    test(
+        'Nested loops with fused assignment and condition edges',
+        testNestedLoopsFusedEdes
+    );
```
[the remark this answers is not in the document]
Comment (phschaad) [reply tests/unit/layouter/state_machine/sm_layouter.test.ts:226]: Added

**Review by tbennun — APPROVED** [review]
LGTM!

**Pull request merged**